hil/src/download_s3.rs

```diff
@@ -113,58 +114,62 @@ pub async fn download_url(
     .wrap_err("failed to create tempfile")?;
 
     let tmp_file: Arc<File> = Arc::new(tmp_file);
-    let bytes_downloaded = Arc::new(AtomicU64::new(0));
-    let out_path_clone = out_path.to_owned();
     let tmp_file_path = Arc::new(tmp_file_path);
+
+    let ranges = Arc::new(Mutex::new(ranges));
     let mut tasks = JoinSet::new();
+    let bytes_downloaded = Arc::new(AtomicU64::new(0));
 
-    for part_number in 0..num_parts {
-        let semaphore = semaphore.clone();
+    for _ in 0..CONCURRENCY {
+        let ranges = Arc::clone(&ranges);
         let client = client.clone();
         let bucket = s3_parts.bucket.clone();
         let key = s3_parts.key.clone();
-        let tmp_file = tmp_file.clone();
+        let tmp_file = Arc::clone(&tmp_file);
         let pb = pb.clone();
-        let bytes_downloaded = bytes_downloaded.clone();
-
-        let start = part_number * part_size;
-        let end = std::cmp::min(start + part_size, bytes_to_download) - 1;
-        let range = ContentRange::new(start, end);
+        let bytes_downloaded = Arc::clone(&bytes_downloaded);
 
         tasks.spawn(async move {
-            let _permit = semaphore.acquire().await;
-
-            let body = download_part_retry_on_timeout(
-                part_number,
-                &range,
-                &client,
-                &bucket,
-                &key,
-            )
-            .await?;
-
-            let chunk_size = body.len() as u64;
-
-            tokio::task::spawn_blocking(move || {
-                tmp_file.write_all_at(&body, start)?;
-                Ok::<(), std::io::Error>(())
-            })
-            .await??;
-
-            if is_interactive {
-                pb.inc(chunk_size);
-            } else {
-                let bytes_so_far = bytes_downloaded
-                    .fetch_add(chunk_size, Ordering::Relaxed)
-                    + chunk_size;
-                let pct = (bytes_so_far * 100) / bytes_to_download;
-                if pct % 5 == 0 {
-                    info!(
-                        "Downloaded: ({}/{} MiB) {}%",
-                        bytes_so_far >> 20,
-                        bytes_to_download >> 20,
-                        pct,
-                    );
+            loop {
+                let range_option = {
+                    let mut ranges_lock = ranges.lock().await;
+                    ranges_lock.next()
+                };
+
+                let range = match range_option {
```

Comment:
prefer let .. else syntax for simple pattern matching where the failure case evaluates to `!` (such as when doing `return` or `break`).